File: backend/code_java/readwordexample/readwordcode/src/main/java/com/code/factory/SubscriberFactory.java
package com.code.factory;

import com.code.countword.Implementation.CountWordImp;
import com.code.countword.Implementation.ReadDirectoryImp;
import com.code.countword.Implementation.ReadFileImp;
import com.code.countword.util.InfoWordCount;

import java.io.File;
import java.util.List;
import java.util.function.Consumer;

public interface SubscriberFactory {

    ReadDirectoryImp getReadDirectory(Consumer<String> callRabbit);

    ReadFileImp getReadFile(File file, InfoWordCount infoWordCount);

    CountWordImp getCountWord(List<String> lines, File file, int task, int numTask, InfoWordCount infoWordCount);
}

File: backend/code_scala/readwordexample/readwordcode/src/main/scala/com/code/factory/SubscriberFactory.scala
package com.code.factory

import com.code.countword.implementation.CountWordImp
import com.code.countword.implementation.ReadDirectoryImp
import com.code.countword.implementation.ReadFileImp
import com.code.countword.util.InfoWordCount

import java.io.File
sealed trait SubscriberFactory {


  def getReadDirectory(callRabbit: String=>Unit): ReadDirectoryImp

  def getReadFile(file: File, infoWordCount: InfoWordCount): ReadFileImp

  def getCountWord(task:Int, numTask:Int, infoWordCount: InfoWordCount): CountWordImp
}

object SubscriberFactory extends SubscriberFactory{
  override def getReadDirectory(callRabbit: String => Unit): ReadDirectoryImp = new ReadDirectoryImp(new InfoWordCount(5,10,callRabbit))

  override def getReadFile(file: File, infoWordCount: InfoWordCount): ReadFileImp = new ReadFileImp(file,infoWordCount)

  override def getCountWord(task: Int, numTask: Int,infoWordCount: InfoWordCount): CountWordImp =
    new CountWordImp(task, numTask, infoWordCount)
}
